MCO-1654: Add runbook for KubeletHealthState (#248)
openshift/runbooks: alerts/machine-config-operator/KubeletHealthState.md

# KubeletHealthState

## Meaning

The Machine Config Daemon (MCD) pods will run a

> health check on kubelet every 30 seconds by running `curl...`

Reviewer: Sorry, I meant this as `curl http://localhost:10248/healthz` as one segment, the ... was due to it splitting lines

Author: Fixed.

# KubeletHealthState

## Meaning

The Machine Config Daemon (MCD) pods will run a

> health check on kubelet every 30 seconds by curling

Reviewer: ```suggestion
health check on kubelet every 30 seconds by running `curl...`
```

Author: Fixed.

`http://localhost:10248/healthz` on the nodes.
If the kubelet is reachable then it will
return `Ok`, if the endpoint does not return this value
then the MCD will increment the `mcd_kubelet_state`

> by 1. The counter will continue to increment so

Reviewer: Maybe consider phrasing like:

```
The counter will continue to increment so
long as the health check does not return an "ok",
and if the counter goes above 2 then the alert will
fire
```

->

```
The counter will continue to increment for each
failed health check, and fire the alert if the value
of `mcd_kubelet_state` > 2.
```

(or is it >=2 ? Probably good to specify)

Author: I did check and it is > then not >= but fixed.

long as the health check does not return an "ok",
and if the counter goes above 2 then the alert will
fire

## Impact

The `kubelet` is a primary node agent that is
required by each node in a cluster to operate.
It facilitates communication, container management,
and is responsible for node administration. If
the kubelet is not running or reachable on a node,

> it will go into `NotReady` status and can become

Reviewer: ```suggestion
the node will go into `NotReady` status and can become
```

Author: Fixed.

inoperable.

## Diagnosis

The health check is initiated and ran by the
machine-config-daemon pod scheduled on each node.
Each pod will log the following when starting up.

```console
I0428 02:23:13.498282   11418 daemon.go:1250] Enabling Kubelet Healthz Monitor
```

The health check will run every 30 seconds,
this check also has a timeout value of 30 seconds

```console
kubeletHealthzPollingInterval = 30 * time.Second
kubeletHealthzTimeout         = 30 * time.Second
```
If the curl does not return `Ok` within that

> 30 seconds or fails to route then it's considered

Reviewer: Could you clarify `fails to route` ? Can consider not specifying that to avoid confusion

Author: Removed that portion since earlier statement should be sufficient.

a failure. The machine-config-daemon pod will
log any failure to `stdout`

```console
W0917 11:21:46.203938 1688435 daemon.go:662] Failed kubelet health check: Get http://localhost:10248/healthz: dial tcp: lookup localhost on <node-ip>:53: no such host
W0917 11:21:46.204070 1688435 daemon.go:596] Got an error from auxiliary tools: kubelet health failure threshold reached
```

Depending on the state of the kubelet the node
may be completely inaccessible via the console

> or `oc` commands. You may need to break into

Reviewer: ```suggestion
or `oc` commands. You may need to login via external console
```

Author: Fixed.

the node via root or SSH directly into the node.

Once in the node you should check the status of
the kubelet service

```console
# sudo systemctl status kubelet.service 
```

Depending on the state of the kubelet it
could be `active`, `inactive` or `failed`.
The command will also display the latest
journal logs from that service. If you
want to see the full journal logs for the
kubelet service you can run:

```console
# journalctl -u kubelet 
```

If the kubelet is failing or inactive the
journal logs are the best place to start
troubleshooting. There could be multiple
reasons the kubelet is failing, including

> other systemd services failing or not

Reviewer: I think for the health monitor to run, the kubelet must have already initialized and ran successfully at some point, such that the daemon is running, so some other service stopping suddenly is not that likely (but I could be wrong here).

Perhaps we can be a bit more vague as to the failure modes, unless we have concrete examples, in which case ignore this comment

Author: This is probably correct since the MCD would rely on the kubelet working in the first place to have a connection to the node. Removed as well as the lists-dependencies command for now.

starting. You can see these dependencies
with the following command:

```console
$ systemctl list-dependencies kubelet.service
```

If the kubelet service is active but the
health check is still failing then, you'll
need to deep dive into why it can't resolve
the curl command.

Check that kubelet is listening on port 10248.

```console
# netstat -antpu | grep 10248
```
You can try restarting kubelet.service or the
underlying node itself.

```console
# systemctl restart kubelet.service
```
```console
# systemctl reboot
```
You should also check the `/etc/hosts` file
and make sure it has the `localhost` and
`localhost.localdomain` entries for both
ipv4 and ipv6

```console
sh-5.1# cat etc/hosts
127.0.0.1   localhost localhost.localdomain localhost4 localhost4.localdomain4
::1         localhost localhost.localdomain localhost6 localhost6.localdomain6
```

## Mitigation

The general state of the node

> effects the state of the kubelet service.

Reviewer: ```suggestion
affects the state of the kubelet service.
```
